src/ZODB/scripts/zodbcmp.py

```diff
@@ -0,0 +1,213 @@
+#                     Kirill Smelkov <kirr@nexedi.com>
+"""Zodbcmp - Tool to compare two ZODB databases
+
+Zodbcmp compares two ZODB databases in between tidmin..tidmax transaction range
+with default range being -∞..+∞ - (whole database).
+
+For comparison both databases are scanned at storage layer and every
+transaction content is compared bit-to-bit between the two. The program stops
+either at first difference found, or when whole requested transaction range is
+scanned with no difference detected.
+
+Exit status is 0 if inputs are the same, 1 if different, 2 if error.
+"""
+
+from __future__ import print_function
+from time import time
+
+def ashex(s):
+    return s.encode('hex')
+
+# something that is greater than everything else
+class Inf:
+    def __cmp__(self, other):
+        return +1
+inf = Inf()
+
+
+# get next item from iter -> (item, !stop)
+def nextitem(it):
+    try:
+        item = it.next()
+    except StopIteration:
+        return None, False
+    else:
+        return item, True
+
+# objects of a IStorageTransactionInformation
+def txnobjv(txn):
+    objv = []
+    for obj in txn:
+        assert obj.tid == txn.tid
+        assert obj.version == ''
+        objv.append(obj)
+
+    objv.sort(key = lambda obj: obj.oid)    # in canonical order
+    return objv
+
+
+# compare two storage transactions
+# 0 - equal, 1 - non-equal
+def txncmp(txn1, txn2):
+    # metadata
+    for attr in ('tid', 'status', 'user', 'description', 'extension'):
+        attr1 = getattr(txn1, attr)
+        attr2 = getattr(txn2, attr)
+        if attr1 != attr2:
+            return 1
+
+    # data
+    objv1 = txnobjv(txn1)
+    objv2 = txnobjv(txn2)
+    if len(objv1) != len(objv2):
+        return 1
+
+    for obj1, obj2 in zip(objv1, objv2):
+        for attr in ('oid', 'data', 'data_txn'):
+            attr1 = getattr(obj1, attr)
+            attr2 = getattr(obj2, attr)
+            if attr1 != attr2:
+                return 1
+
+    return 0
+
+
+# compare two storages
+# 0 - equal, 1 - non-equal
+def storcmp(stor1, stor2, tidmin, tidmax, verbose=False):
+    iter1 = stor1.iterator(tidmin, tidmax)
+    iter2 = stor2.iterator(tidmin, tidmax)
+
+    Tprev = time()
+    txncount = 0
+    while 1:
+        txn1, ok1 = nextitem(iter1)
+        txn2, ok2 = nextitem(iter2)
+
+        # comparison finished
+        if not ok1 and not ok2:
+            if verbose:
+                print("equal")
+            return 0
+
+        # one part has entry not present in another part
+        if txn1 is None or txn2 is None or txn1.tid != txn2.tid:
+            if verbose:
+                tid1 = txn1.tid if txn1 else inf
+                tid2 = txn2.tid if txn2 else inf
+                l = [(tid1, 1,2), (tid2, 2,1)]
+                l.sort()
+                mintid, minstor, maxstor = l[0]
+                print("not-equal: tid %s present in stor%i but not in stor%i" % (
+                        ashex(mintid), minstor, maxstor))
+            return 1
```

Comment:
I've thought about MAX_TID too but:

- there is no MAX_TID in ZODB (it is in NEO) - in ZODB it is `utils.maxtid`
- in both ZODB and NEO it is 7fff... instead of ffff... for signedness. Still having 7f vs ff looks there can be a risk of skipping transactions.
- I wanted code here to be genric, not depending on particular MAX value a tid can be. This can be achieved with `inf` approach.